Manually merge 3.0.x into master

greg0ire opened the pull request:
No conflicts besides `composer.lock` baseline changed.

greg0ire committed:
Setup static analysis with Psalm

greg0ire committed:
Install dependencies compatible with 7.2

This should make it easier to work with Composer from machines using
different versions of PHP than the lowest one we use in the CI.

greg0ire committed:
Merge pull request #3951 from greg0ire/sa-with-psalm

Setup static analysis with Psalm

greg0ire committed:
Merge remote-tracking branch 'origin/2.10.x' into 2.11.x

greg0ire committed:
Merge branch '2.11.x' into 3.0.x

greg0ire committed:
Add missing use statement

greg0ire committed:
Upgrade nikic/php-parser

For some reason a low version crashes Psalm.

greg0ire committed:
Avoid conflict with parent phpdoc

greg0ire committed:
Add stub file for PDO

greg0ire committed:
Mark exceptions as immutable

greg0ire committed:
Suppress errors about SqlAnywhere symbols

greg0ire committed:
Suppress errors about unreleased stubs

greg0ire committed:
Use a proper default value

null contradicts the phpdoc

morozov committed:
Merge pull request #3955 from greg0ire/smaller-baseline

Remove baseline

morozov committed:
Reworked LoggingTest to be able to test Statement::executeUpdate()

greg0ire commented:
@morozov I need help regarding the 2 last error I get. I don't know how I should fix them:

```
ERROR: ImpureMethodCall - src/Exception/DriverException.php:24:57 - Cannot call an possibly-mutating method Doctrine\DBAL\Driver\DriverException::getCode from a mutation-free context (see https://psalm.dev/203)
        parent::__construct($message, $driverException->getCode(), $driverException);


ERROR: ImpureMethodCall - src/Exception/DriverException.php:29:28 - Cannot call an possibly-mutating method Doctrine\DBAL\Exception\DriverException::getPrevious from a mutation-free context (see https://psalm.dev/203)
        $previous = $this->getPrevious();
```

morozov committed:
Merge pull request #3957 from morozov/logging-test

Reworked LoggingTest to be able to test Statement::executeUpdate()

morozov commented:
@greg0ire it looks like the problem is that `Doctrine\DBAL\Driver\DriverException::getCode()` is inherited from `Throwable` which is unknown to Psalm. For instance, declaring the method explicitly fixes one issue:
```php
interface DriverException extends Throwable
{
    /* ... */
    public function getCode();
}
```
Although, I'm not sure if it's the right solution since Psalm itself should be aware that the built-in `Exception` is immutable.

It's fine if we suppress them for now since we're just setting up the analysis.

greg0ire commented:
:thinking: I don't remember why psalm asked me to mark all exceptions as immutable, but when I remove this annotation everywhere, I no longer get the error… this is very confusing.

greg0ire commented:
Looks like this commit I made is at the origin of the chain reaction: 5a61341fea474990fe2c4fb5e27515c053ccf4a4

greg0ire commented:
Ok so it all stems from this message I got on 2.10.x, and that I fixed too hastily:

```
ERROR: MissingImmutableAnnotation - lib/Doctrine/DBAL/Driver/DriverException.php:30:5 - Throwable::getMessage is marked immutable, but Doctrine\DBAL\Driver\DriverException::getMessage is not marked immutable (see https://psalm.dev/213)
    /**
     * Returns the driver error message.
     *
     * @return string
     */
    public function getMessage();
```

I'm going to revert that commit, and just mark the method immutable, not the whole class.

greg0ire committed:
Stop relying on the master version of Psalm

I just got bitten by a bug on master at
https://github.com/doctrine/sql-formatter/pull/43#issuecomment-615576443,
and there are good chances we have the bug on the Dbal too. Anyway, it
only makes sense to use the same version as in the vendors.

greg0ire committed:
Mark every exception as immutable

morozov committed:
Merge pull request #3964 from greg0ire/moar-immutable-exceptions

Mark every exception as immutable

morozov committed:
Merge pull request #3961 from greg0ire/stop-relying-on-psalm-master

Stop relying on the master version of Psalm

greg0ire committed:
Merge remote-tracking branch 'origin/2.10.x' into 2.11.x

greg0ire committed:
Merge branch '2.11.x' into 3.0.x

greg0ire committed:
Merge branch '3.0.x'

greg0ire commented on psalm.xml:
```diff
@@ -0,0 +1,66 @@
+                -->
+                <file name="src/Exception/DriverException.php"/>
+            </errorLevel>
+        </ImpureMethodCall>
```
@morozov I added this

greg0ire committed:
Sync expected coverage report count

See e360ba0 for AppVeyor. For ContinuousPHP, I am not how to track the
history of the pipeline, or if there is one.

greg0ire committed:
Merge pull request #3965 from greg0ire/adjust-expected-coverage-reports-number

Sync expected coverage report count

greg0ire committed:
Merge remote-tracking branch 'origin/3.0.x'

greg0ire commented:
@morozov this one is ready too
